[DPE-6794] - Add cos docs (#57)
canonical/charmed-etcd-operator: docs/monitoring.md

# Monitoring (COS)
The Canonical Observability Stack (COS) is a set of tools that facilitates gathering, processing, visualizing, and setting up alerts on telemetry signals generated by workloads in and outside of Juju.

The etcd charm can use COS to connect to Grafana and Prometheus to use monitoring, alert rules, and log features.

> ## Deploying COS

Reviewer: ```suggestion
## Deploy COS
```
Our style guide encourages us to generally use the imperative in headings (especially when they're about how to perform a specific action)

> Deploy the [`cos-lite`](https://charmhub.io/topics/canonical-observability-stack) bundle in a Kubernetes environment following the [MicroK8s guide](https://charmhub.io/topics/canonical-observability-stack/tutorials/install-microk8s). Since the etcd charm is deployed directly on a cloud infrastructure environment, it is needed to offer the endpoints of the COS relations. The [offers-overlay](https://github.com/canonical/cos-lite-bundle/blob/main/overlays/offers-overlay.yaml) can be used, and this step is shown in the [COS tutorial](https://charmhub.io/topics/canonical-observability-stack/tutorials/install-microk8s#heading--deploy-the-cos-lite-bundle-with-overlays).

Reviewer: ```suggestion
Deploy the [`cos-lite`](https://charmhub.io/topics/canonical-observability-stack) bundle in a Kubernetes environment following the [MicroK8s guide](https://charmhub.io/topics/canonical-observability-stack/tutorials/install-microk8s). Since the etcd charm is deployed directly on a cloud infrastructure environment, it requires offering the endpoints of the COS relations. The [offers-overlay](https://github.com/canonical/cos-lite-bundle/blob/main/overlays/offers-overlay.yaml) can be used, and this step is shown in the [COS tutorial](https://charmhub.io/topics/canonical-observability-stack/tutorials/install-microk8s#heading--deploy-the-cos-lite-bundle-with-overlays).
```

cos    k8s         microk8s/localhost  3.6.4    unsupported  12:26:22Z

App           Version  Status  Scale  Charm             Channel        Rev  Address         Exposed  Message
alertmanager  0.27.0   active      1  alertmanager-k8s  latest/stable  154  10.152.183.216  no
catalogue              active      1  catalogue-k8s     latest/stable   80  10.152.183.174  no
grafana       9.5.3    active      1  grafana-k8s       latest/stable  138  10.152.183.234  no
loki          2.9.6    active      1  loki-k8s          latest/stable  186  10.152.183.31   no
prometheus    2.52.0   active      1  prometheus-k8s    latest/stable  232  10.152.183.55   no
traefik       2.11.0   active      1  traefik-k8s       latest/stable  232  10.152.183.116  no       Serving at 125.121.179.159

Unit             Workload  Agent  Address       Ports  Message
alertmanager/0*  active    idle   10.1.153.187
catalogue/0*     active    idle   10.1.153.177
grafana/0*       active    idle   10.1.153.189
loki/0*          active    idle   10.1.153.188
prometheus/0*    active    idle   10.1.153.190
traefik/0*       active    idle   10.1.153.186         Serving at 125.121.179.159
```

> ## Integrating etcd with COS

Reviewer: ```suggestion
## Integrate etcd with COS
```

### Offer interfaces via the COS controller
Switch to COS environment and offer COS interfaces to be cross-model related with VM model that you are using to deploy the etcd charm:

```shell
juju switch cos

juju offer grafana:grafana-dashboard
juju offer loki:logging
juju offer prometheus:receive-remote-write
```

> ### Consuming the COS offers on the etcd model

Reviewer: ```suggestion
### Consume the COS offers on the etcd model
```

Switch back to the etcd model and run the following commands to consume the COS offers:

```shell

> juju switch cos

Reviewer: Is this a typo? (should the etcd model have another name?)

Author: Yes, that is a typo. Thank you for catching it.

juju consume k8s:admin/cos.prometheus-receive-remote-write
juju consume k8s:admin/cos.loki-logging
juju consume k8s:admin/cos.grafana-dashboards
```

> > In the previous commands, `k8s` is the name of the controller where the COS model is deployed, and `cos` is the name of the model where the COS model is deployed.

Reviewer: This is a bit confusing to read. Would something like this also work?
```suggestion
> In the commands above, `k8s` refers to the controller where the COS-lite bundle is deployed, and `cos` refers to the COS model.
```

> ### Deploying grafana-agent

Reviewer: ```suggestion
### Deploy `grafana-agent`
```

The [Grafana agent](https://charmhub.io/grafana-agent) is a lightweight, open-source agent that runs on your host and sends metrics and logs to Grafana Cloud. The Grafana agent is deployed as a sidecar container in the etcd charm. To deploy the Grafana agent, run the following command:

```shell

> juju deploy grafana-agent --channel stable --base ubuntu@24.04

Reviewer: Is `--channel stable` necessary?

Author: I believe we can make do without it.

prometheus-receive-remote-write  active  k8s    admin/cos.prometheus-receive-remote-write

App            Version  Status  Scale  Charm          Channel      Rev  Exposed  Message
charmed-etcd            active      3  charmed-etcd                  0  no
grafana-agent           active      3  grafana-agent  latest/edge  457  no       tracing: off

Unit                Workload  Agent      Machine  Public address  Ports  Message
charmed-etcd/0*     active    idle       0        10.8.159.214
  grafana-agent/3   active    executing           10.8.159.214           tracing: off
charmed-etcd/1      active    idle       1        10.8.159.42
  grafana-agent/2   active    executing           10.8.159.42            tracing: off
charmed-etcd/2      active    idle       2        10.8.159.235
  grafana-agent/0*  active    executing           10.8.159.235           tracing: off

Machine  State    Address       Inst id        Base          AZ  Message
0        started  10.8.159.214  juju-604e01-0  ubuntu@24.04      Running
1        started  10.8.159.42   juju-604e01-1  ubuntu@24.04      Running
2        started  10.8.159.235  juju-604e01-2  ubuntu@24.04      Running
```

> ### Accessing Grafana dashboard

Reviewer: ```suggestion
### Access Grafana dashboard
```

To get the Grafana dashboard URL, run the following command:

```shell
juju switch k8s

juju run traefik/0 show-proxied-endpoints --format=yaml \
  | yq '."traefik/0".results."proxied-endpoints"' \
  | jq
```

In the output, look for the `grafana` endpoint. The URL should look similar to the following:

```shell
{
  "traefik": {
    "url": "http://125.121.179.159"
  },
  "prometheus/0": {
    "url": "http://125.121.179.159/cos-prometheus-0"
  },
  "loki/0": {
    "url": "http://125.121.179.159/cos-loki-0"
  },
  "alertmanager": {
    "url": "http://125.121.179.159/cos-alertmanager"
  },
  "catalogue": {
    "url": "http://125.121.179.159/cos-catalogue"
  }
}
```

To get the grafana dashboard URL, and admin password, run the following command:

```shell
juju run grafana/leader get-admin-password --model cos
``` 

The output should look similar to the following:

```shell
admin-password: GfPP7OBmvJOe
url: http://125.121.179.159/cos-grafana
```

You can now access the Grafana dashboard using the URL and admin password.

> ### Accessing the etcd dashboard

Reviewer: ```suggestion
### Access the etcd dashboard
```

A default dashboard for etcd is available in Grafana. To access the etcd dashboard, follow these steps:
- Head to the Grafana dashboard URL.
- Log in using the admin username and password.
- Head to the dashboard section and search for the etcd dashboard.

The etcd dashboard should display the metrics collected by the Grafana agent from the etcd charm. It should look similar to the following:

![etcd dashboard](./images/grafana-dashboard.jpg)

> ## The metrics collected by the Grafana agent

Reviewer: ```suggestion
## Metrics collected by the Grafana agent
```

> The metrics exposed by etcd are detailed in the [etcd documentation](https://etcd.io/docs/v3.5/metrics/). 

Reviewer: ```suggestion
The metrics exposed by etcd are detailed in the [upstream etcd documentation](https://etcd.io/docs/v3.5/metrics/). 
```

> ## The default alerts for the etcd charm

Reviewer: ```suggestion
## Default alerts for the etcd charm
```

url: http://125.121.179.159/cos-grafana
```

You can now access the Grafana dashboard using the URL and admin password.

### Accessing the etcd dashboard

A default dashboard for etcd is available in Grafana. To access the etcd dashboard, follow these steps:
- Head to the Grafana dashboard URL.
- Log in using the admin username and password.
- Head to the dashboard section and search for the etcd dashboard.

The etcd dashboard should display the metrics collected by the Grafana agent from the etcd charm. It should look similar to the following:

![etcd dashboard](./images/grafana-dashboard.jpg)

## The metrics collected by the Grafana agent
The metrics exposed by etcd are detailed in the [etcd documentation](https://etcd.io/docs/v3.5/metrics/). 

## The default alerts for the etcd charm
The etcd charm comes with default alerts that are set up in Grafana. These alerts are based on the metrics collected by the Grafana agent. The default alerts are detailed in the [etcd documentation](https://etcd.io/docs/v3.5/op-guide/monitoring/#alerting). 

They include alerts for:
Certainly, here's a breakdown of each alert within the "etcd" group, detailing their purpose and conditions:

> 

Reviewer 2: This doesn't make any sense to me, I think it needs to be fixed.

Author: Fixed. That was my bad. I used Gemini to summarize the alerts before tweaking them and forgot to remove the introduction 🤦.
